### DISCUSSION TITLE
[자성튜] 도대체 GC는 언제 발생할까?!

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cs-collections/cs-collections-discussion#12] [Java]
_**우리는 JVM 영역 중에 Heap 영역만이 GC의 관리 대상이라는 것을 알고 있다.**_

**Heap 메모리**
> 클래스 인스턴스, 배열이 여기에 쌓이며 이 메모리는 공유 메모리라고도 불린다. 여러 스레드에서 공유하는 데이터들이 저장되는 메모리다.

### GC의 원리
GC 작업을 하는 가비지 컬렉터는
- 메모리 할당
- 사용 중인 메모리 인식
- 사용하지 않는 메모리 인식
의 역할을 한다. 

자 이제 GC가 관리하는 Heap 영역안에서 객체들이 생성되고 GC되는 과정에 대해 알아보자.

![image](https://user-images.githubusercontent.com/65801290/208047516-4617062d-f0ee-426c-8a37-98cd360fd83e.png)
힙의 영역이 이렇게 되어있는데

1. 일단 메모리에 객체가 생성되면, Eden 영역에 객체가 생성된다.
2. Eden 영역에 데이터가 꽉 차면, 이 영역에 있던 객체가 없어지거나 옮겨져야한다. 이때 옮기는 위치가 Survivor0 또는 Survivor1의 위치로 간다.
3. 이 때(Eden 영역에서 Survivor 영역으로 넘어갈 때) Minor GC가 발생하며 참조되지 않는 객체는 메모리에서 제거된다.
4. 두 Survivor 의 영역 중 하나는 반드시 비어있어야 하고 그 비어 있는 곳에 Eden 영역에 있던 객체 중 GC로 부터 살아남은 놈들이 가게 된다.
5. 그러면 이제 Eden에 또 새로운 객체를 받을 수 있을 것이고 Survivor 0 또는 1에 객체들이 있을 것이다 나머지 한 곳은 비어 있을 것이다.
6. 그러다 이제 객체들이 있던 Survivor 하나가 꽉 차게 되면 또 Minor GC가 발생하여 Eden 영역에 있던 객체와 꽉 찬 Survivor 영역에 있던 객체들이 비어있던 Survivor 영역으로 이동한다.
7. 이렇게 계속 반복되며 계속 살아남은 객체들은 이제 Old 영역으로 이동하게 된다. 그리고 Eden 영역에 있는 객체의 크기가 큰 것들은 바로 Old 영역으로 넘어간다.
8. 이제 이러다 보면 Old 영역도 꽉 차는 시점이 올텐데 그때 Major GC가 일어난다. Major GC는 Minor GC보다 오랜 시간이 걸린다.

**_Young 영역에서 살아남아 Old 영역에 입장할 수 있는 기준_**
> 각 객체는 Minor GC에서 살아남은 횟수를 기록하는 age bit 를 가지고 있으며, 
> Minor GC가 발생할 때마다 age bit 값은 1씩 증가 하게되며, age bit 값이 MaxTenuringThreshold 라는 설정값을 초과하게 되는 경우 Old 
> Generation 영역에 객체가 이동된다. 또 Young 영역에 Memory가 부족한 경우에도 이동된다.

### Stop The World ! (STW)
Minor GC든 Major GC든 일단 GC가 동작하는 순간 GC를 제외한 모든 쓰레드는 중지된다.
그냥 App Service가 중지 된다는 것.
Young 영역의 Minor GC는 객체 수명이 짧고 많은 객체 검사가 일어나지 않기에 속도가 매우 빨라 애플리케이션에 거의 영향이 없다.
그러나 Major GC는 모든 객체를 검사해야해서 시간이 오래걸린다 그만큼 App 중지 시간도 오래 걸리는 것. 
그래서 Major GC는 최소화 하려고 노력해야한다.

JDK 7이상에서 지원하는 GC 방식에는 5가지가 있으며
- 시리얼 콜렉터
- 병렬 콜렉터
- 병렬 콤팩팅 콜렉터
- CMS 콜렉터 
- G1 콜렉터
이 GC들은 WAS나 JAVA App 수행시 옵션을 지정하여 선택 할 수 있다.

우리는 GC를 튜닝하여 성능을 좋게 해보자 하며 까불면 안된다.
GC 튜닝도 튜닝이지만 그러기 앞서서 객체 생성을 최대한 줄여야 한다.

우리는 현실적으로 코드에서 GC를 줄이는 방법들이 있다.
- StringBuilder, StringBuffer 생활화
- Log 줄이기
- 불변 객체 사용하기.
- 컬렉션 배열 등을 사용할 때 capacity 정해주기.
- 10~100 MB짜리 XML이나 JSON를 파싱하면 엄청난 임시 메모리를 사용하는데 어렵겠지만 JSON, XML 파싱 줄이기..

**Comment by Irisation23** [comment]
> 8. 이제 이러다 보면 Old 영역도 꽉 차는 시점이 올텐데 그때 Major GC가 일어난다. Major GC는 `Mino` GC보다 오랜 시간이 걸린다.

- 해당 부분 `Mino` -> `Minor` 로 수정이 필요해 보입니다!

**Reply by ahngilwoong** [reply]
감사합니다 수정했습니다.
